lib/Reconciler.lua

```diff
@@ -1,542 +1,361 @@
-	was created.
-]]
-function Reconciler._setRbxProp(rbx, key, value, element)
-	if type(key) == "string" then
-		-- Regular property
+local function mountTree(element, parentRbx, key)
+	assert(Type.is(element, Type.Element))
+	assert(typeof(parentRbx) == "Instance" or typeof(parentRbx) == "nil")
+
+	local tree = {
+		[Type] = Type.Tree,
+		tasks = {},
+		taskIndex = 1,
+		tasksRunning = false,
+		connections = {},
+		mounted = true,
+		rootNode = nil,
+	}
+
+	if ASYNC_SCHEDULER then
+		tree.renderStepId = getGuid()
+		RunService:BindToRenderStep(tree.renderStepId, Enum.RenderPriority.Last.Value, function()
+			processTreeTasksAsync(tree, ASYNC_BUDGET_PER_FRAME)
+		end)
+	end
 
-		local success, err = pcall(set, rbx, key, value)
+	scheduleTask(tree, taskMountNode({
+		element = element,
+		key = key,
+		parentNode = nil,
+		parentRbx = parentRbx,
+		isTreeRoot = true,
+	}))
 
-		if not success then
-			local source = element.source or DEFAULT_SOURCE
+	return tree
+end
 
-			local message = ("Failed to set property %s on primitive instance of class %s\n%s\n%s"):format(
-				key,
-				rbx.ClassName,
-				err,
-				source
-			)
+local function unmountTree(tree)
+	assert(Type.is(tree, Type.Tree))
+	assert(tree.mounted, "not mounted")
 
-			error(message, 0)
-		end
-	elseif type(key) == "table" then
-		-- Special property with extra data attached.
+	tree.mounted = false
 
-		if key.type == Event then
-			Reconciler._singleEventManager:connect(rbx, key.name, value)
-		elseif key.type == Change then
-			Reconciler._singleEventManager:connectProperty(rbx, key.name, value)
-		else
-			local source = element.source or DEFAULT_SOURCE
+	if ASYNC_SCHEDULER then
+		RunService:UnbindFromRenderStep(tree.renderStepId)
+	end
 
-			-- luacheck: ignore 6
-			local message = ("Failed to set special property on primitive instance of class %s\nInvalid special property type %q\n%s"):format(
-				rbx.ClassName,
-				tostring(key.type),
-				source
-			)
+	-- TODO: Flush/cancel existing tasks and unmount asynchronously?
 
-			error(message, 0)
-		end
-	elseif type(key) ~= "userdata" then
-		-- Userdata values are special markers, usually created by Symbol
-		-- They have no data attached other than being unique keys
+	scheduleTask(tree, taskUnmountNode({
+		node = tree.rootNode,
+	}))
 
-		local source = element.source or DEFAULT_SOURCE
+	-- For now, flush the entire tree and unmount synchronously
+	processTreeTasksSync(tree)
+end
 
-		local message = ("Properties with a key type of %q are not supported\n%s"):format(
-			type(key),
-			source
-		)
+local function reconcileTree(tree, toElement)
+	assert(Type.is(tree, Type.Tree))
+	assert(Type.is(toElement, Type.Element))
 
-		error(message, 0)
-	end
+	scheduleTask(tree, taskReconcileNode({
+		node = tree.rootNode,
+		toElement = toElement,
+	}))
 end
 
-return Reconciler
+return {
+	mount = mountTree,
+	unmount = unmountTree,
+	reconcileTree = reconcileTree
```

Comment:
tree is removed in the other two function names, but not reconcileTree. mount takes a node as an argument but for some reason its function is mountTree. unmount actually takes a tree, so I don't completely understand why tree was removed from its name. I'd almost say drop the "tree" in all these function names.

Reply:
This is a debug API, not entirely concerned with the outward semantics except that I can run example code against it!